### PR TITLE
[profiler] Implement global flush points for thread-local buffers.

### PR DESCRIPTION
**Pull request by alexrp**
Plus a number of other improvements to the profiler as well as minor fixes in `mono/utils` code.

I would strongly recommend reviewing commits one-by-one rather than looking at the full diff.

### CODE REVIEW

**Comment by kumpera** [comment]
We should remove the free func code from lls, its not used by the runtime. it's a timebomb before someone uses it and stuff breaks.

File: mono/profiler/proflog.c [hunk 1]
```diff
@@ -526,31 +541,36 @@ typedef struct {
 	uint64_t time;
 } MethodInfo;
 
-#ifdef TLS_INIT
-#undef TLS_INIT
-#endif
-
 #ifdef HOST_WIN32
```
Comment (kumpera) [inline mono/profiler/proflog.c]: we should really kill this stuff and have the runtime provide tls for the profiler

File: mono/profiler/proflog.c [hunk 2]
```diff
@@ -3894,11 +3995,16 @@ log_shutdown (MonoProfiler *prof)
...
-		send_buffer (prof, TLS_GET (GPtrArray, tlsmethodlist), TLS_GET (LogBuffer, tlsbuffer));
-
-	TLS_SET (tlsbuffer, NULL);
-	TLS_SET (tlsmethodlist, NULL);
+	/*
+	 * Ensure that we empty the LLS completely, even if some nodes are
+	 * not immediately removed upon calling mono_lls_remove (), by
+	 * iterating until the head is NULL.
+	 */
+	while (profiler_thread_list.head) {
```
Comment (kumpera) [inline mono/profiler/proflog.c]: wont this cause other threads to access invalid memory?

File: mono/profiler/proflog.c [hunk 3]
```diff
@@ -487,8 +488,102 @@ ign_res (int G_GNUC_UNUSED unused, ...)
...
+static void
+buffer_unlock (void)
+{
+	mono_memory_barrier ();
+
+	// See the comment in buffer_lock ().
+	if (InterlockedReadPointer (&buffer_rwlock_exclusive) == (gpointer) thread_id ())
+		return;
+
+	g_assert (InterlockedRead (&buffer_rwlock_count) && "Why are we trying to decrement a zero reader count?");
```
Comment (kumpera) [inline mono/profiler/proflog.c]: && a string??

File: mono/profiler/proflog.c [hunk 4]
```diff
@@ -487,8 +488,102 @@ ign_res (int G_GNUC_UNUSED unused, ...)
...
+	 * In other words, we rely on the fact that the GC thread takes
+	 * the exclusive lock in the gc_event () callback when the world
+	 * is about to stop.
+	 */
+	if (excl == tid)
+		goto done;
+
+	MONO_ENTER_GC_SAFE;
+
+	while ((excl = InterlockedReadPointer (&buffer_rwlock_exclusive)) && excl != tid)
```
Comment (kumpera) [inline mono/profiler/proflog.c]: how would the second part of the cond ever be false?

File: mono/profiler/proflog.c [hunk 5]
```diff
@@ -487,8 +488,102 @@ ign_res (int G_GNUC_UNUSED unused, ...)
...
+		mono_thread_info_yield ();
+
+	InterlockedIncrement (&buffer_rwlock_count);
```
Comment (kumpera) [inline mono/profiler/proflog.c]: if excl == tid this would happen and leave it unbalanced compared to line 526

File: mono/profiler/proflog.c (hunk 3 above)
Comment (evincarofautumn) [inline mono/profiler/proflog.c]: It’s an old-fashioned way to include an assertion message, like:

```
assert(!"This should never happen.");
```

File: mono/profiler/proflog.c [hunk 6]
```diff
@@ -1104,85 +1225,145 @@ static uint64_t last_hs_time = 0;
...
 			gc_count++;
-	}
-	if (ev == MONO_GC_EVENT_PRE_START_WORLD)
+		break;
+	case MONO_GC_EVENT_PRE_STOP_WORLD:
+		/*
+		 * Ensure that no thread can be in the middle of writing to
+		 * a buffer when the world stops...
+		 */
+		buffer_lock_excl ();
```
Comment (kumpera) [inline mono/profiler/proflog.c]: what about one off suspends?

File: mono/profiler/proflog.c [hunk 7]
```diff
@@ -487,8 +488,102 @@ ign_res (int G_GNUC_UNUSED unused, ...)
...
+static void
+buffer_lock_excl (void)
+{
+	gpointer tid = (gpointer) thread_id ();
+
+	g_assert (InterlockedReadPointer (&buffer_rwlock_exclusive) != tid && "Why are we taking the exclusive lock twice?");
+
+	MONO_ENTER_GC_SAFE;
+
+	while (InterlockedCompareExchangePointer (&buffer_rwlock_exclusive, tid, 0))
```
Comment (kumpera) [inline mono/profiler/proflog.c]: so you went with a dekker lock. definitely better. it bias exclusive over shared, which is what we want.

I think it's correct.

File: mono/profiler/proflog.c [hunk 8]
```diff
@@ -959,16 +1074,14 @@ remove_thread (MonoProfiler *prof, MonoProfilerThread *thread, gboolean from_cal
 
 			uint64_t now = current_time ();
 
-			ENTER_LOG (buffer, "thread-end");
```
Comment (kumpera) [inline mono/profiler/proflog.c]: we're losing this tracing capability. have you ever used it?

**Comment by kumpera** [comment]
few things.

- we should not duplicate thread info management in proflog. adding an api to the runtime would allow us to surface a smaller and higher level API.

- there's one bit on flushing thread events on shutdown that Im not sure on correctness.

- this removes implicit flushes of thread local buffers. I wonder whether those were not hiding existing ordering bugs.

- the blocking locking in stw start and one off suspends worry me.

**Comment by kumpera** [comment]
p1 can be done layer
p2 and p4 need to be addressed
p3 could use the mlpd linter I started to verify them

finally. sync events are not generating a sync event that can be used to narrow uncertainty windows.

File: mono/profiler/proflog.c (hunk 3 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: It's a style commonly used in LLVM that I decided to use to make the assertions a bit more meaningful.

File: mono/profiler/proflog.c (hunk 2 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: You might be right. I'd forgotten about threads like the finalizer thread and TP threads. Will have a closer look.

File: mono/profiler/proflog.c (hunk 6 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: Not entirely sure what you mean by that. Let's discuss this one Skype?

File: mono/profiler/proflog.c (hunk 4 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: It can't, this looks like a leftover from an older (non-working) design of the lock.

File: mono/profiler/proflog.c (hunk 2 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: OK, I did some digging, and this should not be an issue. By the time this shutdown callback is invoked, *all* threads have been shut down (including finalizer and TP threads), and the profiler event flags are set to zero immediately after this callback returns. So no more events will come from the main thread, either.

File: mono/profiler/proflog.c (hunk 1 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: We could do that, but perhaps it's a good idea to wait until we kill off `__thread` support (if we do) to reduce complexity.

File: mono/profiler/proflog.c (hunk 8 above)
Comment (alexrp) [inline mono/profiler/proflog.c]: It was there to catch things like:

```c
static void
some_callback (MonoProfiler *prof)
{
    ENTER_LOG;

    LogBuffer *buf = ensure_logbuf (...);

    /* write some data to buf */

    /* get some data from the runtime, triggers some_other_callback */
    stuff = get_something_from_runtime ();

    /* write stuff to buf */

    EXIT_LOG;
}

static void
some_other_callback (MonoProfiler *prof)
{
    ENTER_LOG;

    /* here we'd end up writing in the middle of another event */
}
```

The correct way to write the above would be to call `get_something_from_runtime` outside of the `ENTER_LOG`/`EXIT_LOG` pair.

I've never seen this catch any actual errors in the profiler, though, and I've gone through and checked all event callbacks to make sure they don't call runtime functions while writing to a buffer. Unfortunately, even if this debugging code would have caught problems, all it would do is write to `stderr`, not assert or anything. So there's a good chance it would just be missed.

That said, the reason I removed this debugging code was that I originally intended to remove the `ENTER_LOG`/`EXIT_LOG` macros. But as you can see, I settled on keeping them to create a scope around buffer writing. So, we could just replace the old debugging code with some code that actually asserts if such a case is ever hit.

File: mono/profiler/proflog.c (hunk 6 above)
Comment (kumpera) [inline mono/profiler/proflog.c]: threads can be suspended outside of STW, if a thread is suspended while holding the shared lock we can run into trouble. In theory, at least.

File: mono/profiler/proflog.c (hunk 1 above)
Comment (kumpera) [inline mono/profiler/proflog.c]: Those are unrelated changes. And __thread is not going away any time soon either.

**Comment by alexrp** [comment]
Rebased with the changes we discussed. Please have a look at these new commits:

* https://github.com/mono/mono/pull/3350/commits/886a2e3eaf861a352a77523e5596886440d6dea6
* https://github.com/mono/mono/pull/3350/commits/aebc76fb19fd441971e5319160812c79ed861c80
* https://github.com/mono/mono/pull/3350/commits/a401dc99987dfc7298d4b1d4fe9f3d8c51da54d3

File: mono/metadata/profiler.h [hunk 9]
```diff
@@ -49,7 +50,9 @@ typedef enum {
 	MONO_GC_EVENT_PRE_STOP_WORLD,
 	MONO_GC_EVENT_POST_STOP_WORLD,
 	MONO_GC_EVENT_PRE_START_WORLD,
-	MONO_GC_EVENT_POST_START_WORLD
+	MONO_GC_EVENT_POST_START_WORLD,
+	MONO_GC_EVENT_PRE_STOP_WORLD_LOCKED,
```
Comment (kumpera) [inline mono/metadata/profiler.h]: We need to somewhere explain those events.

**Comment by kumpera** [comment]
LGTM

**Comment by alexrp** [comment]
Rebased with two more commits:

* https://github.com/mono/mono/pull/3350/commits/8f8b8527d8c3bcbff642f64bf4e3a55891a120fb
* https://github.com/mono/mono/pull/3350/commits/d9c84f25b041c38619fa97a4383ab837b3c2896b

@kumpera if these look OK to you, I'd say this is finally mergeable.

**Comment by alexrp** [comment]
Opening a new combined PR for all of my profiler changes to avoid merge conflicts.
